### ISSUE TITLE
Unclear error when implementing both `From<T> for U` and `TryFrom<T> for U`

### ISSUE THREAD

**Issue by Urhengulas** [rust-lang/rust#98062]
Given the following code ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e3aa90a7aa73e047e643f2a72c79add5)): 

```rust
struct S {}

impl From<()> for S {
    fn from(value: ()) -> Self {
        todo!()
    }
}

impl TryFrom<()> for S {
    type Error = ();

    fn try_from(value: ()) -> Result<Self, Self::Error> {
        todo!()
    }
}

fn main() {}
```

The current output is:

```
error[E0119]: conflicting implementations of trait `std::convert::TryFrom<()>` for type `S`
 --> src/main.rs:9:1
  |
9 | impl TryFrom<()> for S {
  | ^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: conflicting implementation in crate `core`:
          - impl<T, U> TryFrom<U> for T
            where U: Into<T>;

For more information about this error, try `rustc --explain E0119`.
error: could not compile `krate` due to previous error
```

Ideally the output should look like:

```
error[E0119]: conflicting implementations of trait `std::convert::TryFrom<()>` for type `S`
 --> src/main.rs:9:1
  |
5 | impl From<()> for S {
  | --------------------- first implementation here
9 | impl TryFrom<()> for S {
  | ^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `S`
  |
  = note: conflicting implementation in crate `core`:
          - impl<T, U> TryFrom<U> for T
            where U: Into<T>;

For more information about this error, try `rustc --explain E0119`.
error: could not compile `krate` due to previous error
```

Me and a colleague hit this problem while refactoring some code from `From` to `TryFrom`. We kept the `From` implementation around and additionally implemented `TryFrom`. It took us a bit of time to figure out the cause of that problem.

If you implement `From<T> for U` it auto implements `Into<U> for T` and therefore it auto implements `TryFrom<T> for U`. This is convenient, but it also makes it impossible to implement both `From` and `TryFrom` for one conversion. This kinda makes sense, but the error message could be a bit clearer about the reason for the conflict.

The proposed output is not optimal, but I'd like the error message to point at the `impl From` somehow.

**Comment by lyming2007** [comment]
after adding  `use std::convert::TryFrom;` in the beginning of the file, the error was reproduced

**Comment by Urhengulas** [comment]
> after adding `use std::convert::TryFrom;` in the beginning of the file, the error was reproduced

Yes, while `TryFrom` is included in the prelude of the 2021 edition, in the 2018 edition you need to import it.

**Comment by lyming2007** [comment]
how about the output like this?
```
error[E0119]: conflicting implementations of trait `std::convert::TryFrom<()>` for type `S`
   --> main.rs:10:1
    |
10  |   impl TryFrom<()> for S {
    |   ^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `S`
    |
   ::: /home/lym/rust/rust/library/core/src/convert/mod.rs:598:1
    |
598 | / impl<T, U> const TryFrom<U> for T
599 | | where
600 | |     U: ~const Into<T>,
601 | | {
...   |
606 | |     }
607 | | }
    | |_- first implementation here
    |
    = note: conflicting implementation in crate `core`:
            - impl<T, U> TryFrom<U> for T
              where U: Into<T>;

```

**Comment by lyming2007** [comment]
I'm wondering why    `impl TryFrom<()> for S` in line 10 is conflicting with this implementation:
```
::: /home/lym/rust/rust/library/core/src/convert/mod.rs:598:1
    |
598 | / impl<T, U> const TryFrom<U> for T
599 | | where
600 | |     U: ~const Into<T>,
601 | | {
...   |
606 | |     }
607 | | }
```
but if I don't implemen`t impl From<()> for S {`, the conflict gone ?

**Comment by lyming2007** [comment]
> implement `From<T> for U` it auto implements `Into<U> for T` and therefore it auto implements `TryFrom<T> for U`

now I figured it out. It's because the implementation of `From<T> for U `it auto implements `Into<U> for T` and therefore it auto implements `TryFrom<T> for U`. So to fix this(emit the implementation of trait From) is to find out the automatic implementations trigged by the  the implementation of `From<T> for U ` which seems hard

**Comment by Urhengulas** [comment]
> So to fix this(emit the implementation of trait From) is to find out the automatic implementations trigged by the the implementation of `From<T> for U ` which seems hard

That's what I think as well... 🤔 

But maybe there is a less optimal solution which is still better than the current one?

**Comment by estebank** [comment]
@lcnr pinging you because you're working on the new solver.

What's happening here is that we don't keep track of the obligations that were evaluated and accepted when comparing two traits in `rustc_trait_selection::traits::coherence::overlap`. We would need to change this in order to properly point at the `impl From<()> for S`.
